teamcity/user.go

```diff
@@ -0,0 +1,212 @@
+	Items []Group `json:"group,omitempty" xml:"groups"`
+}
+
+// UserList contains list of users
+type UserList struct {
+	Count int    `json:"count,omitempty" xml:"count"`
+	Items []User `json:"user,omitempty" xml:"user"`
+}
+
+// NewUser returns an instance of a User. A non-empty Username, Name and Email is required.
+func NewUser(username string, name string, email string) (*User, error) {
+	if username == "" {
+		return nil, fmt.Errorf("Key is required")
+	}
+
+	if name == "" {
+		return nil, fmt.Errorf("Name is required")
+	}
+
+	if email == "" {
+		return nil, fmt.Errorf("Email is required")
+	}
+
+	return &User{
+		Username: username,
+		Name:     name,
+		Email:    email,
+	}, nil
+}
+
+// UserService has operations for handling Users
+type UserService struct {
+	sling      *sling.Sling
+	httpClient *http.Client
+	restHelper *restHelper
+}
+
+func newUserService(base *sling.Sling, httpClient *http.Client) *UserService {
+	sling := base.Path("users/")
+	return &UserService{
+		httpClient: httpClient,
+		sling:      sling,
+		restHelper: newRestHelperWithSling(httpClient, sling),
+	}
+}
+
+// Create - Creates a new User
+func (s *UserService) Create(user *User) (*User, error) {
+	var created User
+	err := s.restHelper.post("", user, &created, "User")
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &created, nil
+}
+
+// GetByID - Get a User by its User ID
+func (s *UserService) GetByID(ID int) (*User, error) {
+	return s.getByLocator(LocatorID(fmt.Sprint(ID)))
+}
+
+// GetByUsername - Get a User by its User Username
+func (s *UserService) GetByUsername(username string) (*User, error) {
+	return s.getByLocator(LocatorUsername(username))
+}
+
+// GetByName - Get a User by its User Name
+func (s *UserService) GetByName(name string) (*User, error) {
+	return s.getByLocator(LocatorName(name))
+}
+
+func (s *UserService) getByLocator(locator Locator) (*User, error) {
+	var out User
+	err := s.restHelper.get(locator.String(), &out, "User")
+	if err != nil {
+		return nil, err
+	}
+
+	return &out, err
+}
+
+// DeleteByID - Deletes a User by its User ID
+func (s *UserService) DeleteByID(id int) error {
+	return s.deleteByLocator(LocatorID(fmt.Sprint(id)))
+}
+
+// DeleteByName - Deletes a User by its User Name
+func (s *UserService) DeleteByName(name string) error {
+	return s.deleteByLocator(LocatorName(name))
+}
+
+// DeleteByUsername - Deletes a User by its User Username
+func (s *UserService) DeleteByUsername(username string) error {
+	return s.deleteByLocator(LocatorUsername(username))
+}
+
+func (s *UserService) deleteByLocator(locator Locator) error {
+	err := s.restHelper.delete(locator.String(), "User")
+	return err
+}
+
+// List - Get list of all User
+func (s *UserService) List() (*UserList, error) {
+	var out UserList
+	err := s.restHelper.get("", &out, "Users")
+	if err != nil {
+		return nil, err
+	}
+	return &out, err
+}
+
+// GroupAddByID - Add User with userID to Group with groupKey
```

Comment:
This API is an interesting point about the N:N relations. 
I believe on TeamCity's API side these resources are managed on the `/user/groups/...`, but wouldn't it make more sense (or be more intuitive) to have these operations on the "Group" side of things? That is, if we had to choose one side to manage it.

I'll take a closer look at how the API behaves and maybe we can sketch out a design here on this PR from the result of this discovery, wdyt?

Reply:
When considering the problem with `List()`, an idea appeared instead of using the Groups field - to replace it with the `func GetGroupListBy...(userLocator locator, groupKey string, start, count, int)`, similarly for groups the ability to get a list of group members `func GetMemberListBy...(groupLocator locator, userLocator locator, start, count int)`, similar for `Roles`  and  `Properites` , wdyt?

Reply:
Regarding which resource should manage the addition and removal, unfortunately, I only found this [method](https://teamcity-support.jetbrains.com/hc/en-us/community/posts/360000487520-TeamCity-REST-API-Add%20-user-to-group-)

Reply:
> Regarding which resource should manage the addition and removal, unfortunately, I only found this [method](https://teamcity-support.jetbrains.com/hc/en-us/community/posts/360000487520-TeamCity-REST-API-Add%20-user-to-group-)

Regardless of the underlying raw API of TeamCity, the general design idea of this client is to provide a user-friendly, albeit programmatic interface for that interaction. Most of this driving factor comes from the usage by the Terraform Provider, where if we simply translate the TeamCity API to client operations 1:1, the ultimate caller code gets very clunky.

IMO it makes more sense to either:
- Add users to groups via the `GroupService` instead of `UserService`. Somehow I think it's way more intuitive and discoverable to perform operations these operations on the "group" than on the "user", if we have to choose one. From the TeamCity UI, we can do both ways, however.
- Create a third service, `UserGroupMembershipService` and do these explicit membership operations there.